File: desktop/workspaces/angular-app/src/app/services/user.service.ts
/**
 * @license
 * Copyright (c) The Authors (see the AUTHORS file)
 *
 * This source code is licensed under the GNU Affero General Public License v3.0 (AGPLv3) for personal, non-commercial use.
 * You may obtain a copy of the AGPL v3.0 at https://www.gnu.org/licenses/agpl-3.0.html.
 *
 * For commercial use, a separate license must be obtained by purchasing from The Authors.
 * For commercial licensing inquiries, please contact The Authors listed in the AUTHORS file.
 */
import { Injectable } from '@angular/core';
import { LocaltronService } from './localtron.service';
import { CookieService } from 'ngx-cookie-service';
import { ReplaySubject } from 'rxjs';
import { Router } from '@angular/router';

@Injectable({
	providedIn: 'root',
})
export class UserService {
	private token: string = '';

	private userSubject = new ReplaySubject<User>(1);
	public user$ = this.userSubject.asObservable();

	constructor(
		private localtron: LocaltronService,
		private cookieService: CookieService,
		private router: Router
	) {
		this.init();
	}

	async init() {
		this.getToken();
		if (!this.hasToken()) {
			let rsp = await this.login('singulatron', 'changeme');
			this.setToken(rsp.token.token as string);
			if (!this.hasToken()) {
				console.error('Something is wrong with the setting of cookies');
				this.router.navigateByUrl('/login');
				return;
			}
		}

		try {
			let rsp = await this.readUserByToken(this.token);
			this.userSubject.next(rsp.user);
		} catch {
			console.error('Cannot read user even with a token');
			this.router.navigateByUrl('/login');
		}
	}

	getToken(): string {
		if (this.token) {
			return this.token;
		}
		this.token = this.cookieService.get('the_token');
		return this.token;
	}

	setToken(token: string) {
		this.token = token;
		this.cookieService.set('the_token', this.token, 3650, '/', '', true);
	}

	hasToken(): boolean {
		let t = this.cookieService.get('the_token');
		return !!t;
	}

	login(email: string, password: string): Promise<LoginResponse> {
		return this.localtron.call('/user/login', {
			email: email,
			password: password,
		});
	}

	readUserByToken(token: string): Promise<ReadUserByTokenResponse> {
		return this.localtron.call('/user/read-user-by-token', {
			token: token,
		});
	}

	getUsers(): Promise<GetUsersResponse> {
		return this.localtron.call('/user/get-users', {});
	}

	/** Save profile on behalf of a user */
	saveProfile(email: string, name: string): Promise<SaveProfileResponse> {
		let req: SaveProfileRequest = {
			email: email,
			name: name,
		};
		return this.localtron.call('/user/save-profile', req);
	}

	changePassword(
		email: string,
		currentPassword: string,
		newPassword: string
	): Promise<ChangePasswordResponse> {
		let req: ChangePasswordRequest = {
			email: email,
			currentPassword: currentPassword,
			newPassword: newPassword,
		};
		return this.localtron.call('/user/change-password', req);
	}

	/** Create a user - alternative to registration
	 */
	createUser(
		user: User,
		password: string,
		roleIds: string[]
	): Promise<CreateUserResponse> {
		let req: CreateUserRequest = {
			user: user,
			password: password,
			roleIds: roleIds,
		};
		return this.localtron.call('/user/create-user', req);
	}

	getRoles(): Promise<GetRolesResposne> {
		return this.localtron.call('/user/get-roles', {});
	}

	getPermissions(): Promise<GetPermissionsResposne> {
		return this.localtron.call('/user/get-permissions', {});
	}

	setRolePermissions(
		roleId: string,
		permissionIds: string[]
	): Promise<SetRolePermissionsResponse> {
		let req: SetRolePermissionsRequest = {
			roleId: roleId,
			permissionIds: permissionIds,
		};
		return this.localtron.call('/user/set-role-permissions', req);
	}

	deleteRole(roleId: string): Promise<DeleteRoleResponse> {
		let req: DeleteRoleRequest = {
			roleId: roleId,
		};
		return this.localtron.call('/user/delete-role', req);
	}

	deleteUser(userId: string): Promise<DeleteUserResponse> {
		let req: DeleteUserRequest = {
			userId: userId,
		};
		return this.localtron.call('/user/delete-user', req);
	}
}

export interface User {
	id?: string;
	createdAt?: Date;
	updatedAt?: Date;
	deletedAt?: Date | null;
	name?: string;
	email?: string;
	passwordHash?: string;
	roleIds?: string[];
	authTokenIds?: string[];
}

export interface RegisterRequest {
	name?: string;
	email?: string;
}

export interface RegisterResponse {}

export interface LoginRequest {
	name?: string;
	email?: string;
}

export interface LoginResponse {
	token: AuthToken;
}

export interface SaveProfileRequest {
	name?: string;
	email?: string;
}

export interface SaveProfileResponse {}

export interface ChangePasswordRequest {
	email?: string;
	currentPassword?: string;
	newPassword?: string;
}

export interface ChangePasswordResponse {}

export const RoleAdmin: Role = {
	id: 'role.admin',
	name: 'Admin Role',
	permissionIds: [],
};

export const RoleUser: Role = {
	id: 'role.user',
	name: 'User Role',
	permissionIds: [],
};

export interface Role {
	id?: string;
	createdAt?: Date;
	updatedAt?: Date;
	name: string;
	description?: string;
	permissionIds: string[];
}

export interface CreateRoleRequest {
	name: string;
	description: string;
	permissionIds: string[];
}

export interface CreateRoleResponse {
	role?: Role;
}

export interface DeleteRoleRequest {
	roleId?: string;
}

export interface DeleteRoleResponse {}

export interface RemoveRoleRequest {
	userId?: string;
	roleId?: string;
}

export interface RemoveRoleResponse {}

export interface Permission {
	id?: string;
	createdAt?: Date;
	updatedAt?: Date;
	name: string;
	description: string;
}

export interface CreatePermissionRequest {
	name: string;
	description: string;
}

export interface CreatePermissionResponse {}

export interface AuthToken {
	id?: string;
	createdAt?: Date;
	updatedAt?: Date;
	deletedAt?: Date | null;
	userId?: string;
	token?: string;
}

export interface ReadUserByTokenResponse {
	user: User;
}

export interface GetUsersRequest {}

export interface GetUsersResponse {
	users: User[];
}

export interface CreateUserRequest {
	user: User;
	password: string;
	roleIds: string[];
}

export interface CreateUserResponse {}

export interface GetRolesRequest {}

export interface GetRolesResposne {
	roles: Role[];
}

export interface GetPermissionsRequest {}

export interface GetPermissionsResposne {
	permissions: Permission[];
}

export interface SetRolePermissionsRequest {}
export interface SetRolePermissionsResponse {}

export interface DeleteUserRequest {
	userId: string;
}
export interface DeleteUserResponse {}

File: localtron/services/user/user_service.go
/**
 * @license
 * Copyright (c) The Authors (see the AUTHORS file)
 *
 * This source code is licensed under the GNU Affero General Public License v3.0 (AGPLv3).
 * You may obtain a copy of the AGPL v3.0 at https://www.gnu.org/licenses/agpl-3.0.html.
 */
package userservice

import (
	"crypto/rsa"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/singulatron/singulatron/localtron/datastore"
	"github.com/singulatron/singulatron/localtron/logger"
	"github.com/singulatron/singulatron/localtron/router"

	usertypes "github.com/singulatron/singulatron/localtron/services/user/types"
)

type UserService struct {
	router *router.Router

	usersStore       datastore.DataStore
	rolesStore       datastore.DataStore
	permissionsStore datastore.DataStore
	credentialsStore datastore.DataStore
	authTokensStore  datastore.DataStore
	keyPairsStore    datastore.DataStore

	privateKey    *rsa.PrivateKey
	publicKeyPem  string
	serviceUserId string
}

func NewUserService(
	router *router.Router,
	datastoreFactory func(tableName string, instance any) (datastore.DataStore, error),
) (*UserService, error) {
	usersStore, err := datastoreFactory("users", &usertypes.User{})
	if err != nil {
		return nil, err
	}
	rolesStore, err := datastoreFactory("roles", &usertypes.Role{})
	if err != nil {
		return nil, err
	}
	authTokensStore, err := datastoreFactory("authTokens", &usertypes.AuthToken{})
	if err != nil {
		return nil, err
	}
	permissionsStore, err := datastoreFactory("permissions", &usertypes.Permission{})
	if err != nil {
		return nil, err
	}
	credentialsStore, err := datastoreFactory("user_credetentials", &usertypes.Credential{})
	if err != nil {
		return nil, err
	}
	keyPairsStore, err := datastoreFactory("keyPairs", &usertypes.KeyPair{})
	if err != nil {
		return nil, err
	}

	service := &UserService{
		router:           router,
		usersStore:       usersStore,
		rolesStore:       rolesStore,
		authTokensStore:  authTokensStore,
		permissionsStore: permissionsStore,
		credentialsStore: credentialsStore,
		keyPairsStore:    keyPairsStore,
	}

	err = service.registerRoles()
	if err != nil {
		return nil, err
	}

	err = service.registerPermissions()
	if err != nil {
		return nil, err
	}

	err = service.bootstrap()
	if err != nil {
		return nil, err
	}

	return service, nil
}

func (s *UserService) bootstrap() error {
	// bootstrapping keys

	keyPairs, err := s.keyPairsStore.Query(
		datastore.All(),
	).Find()
	if err != nil {
		return err
	}

	if len(keyPairs) > 0 {
		kp := keyPairs[0].(*usertypes.KeyPair)
		privKey, err := privateKeyFromString(kp.PrivateKey)
		if err != nil {
			return err
		}
		s.privateKey = privKey
		s.publicKeyPem = kp.PublicKey
	} else {
		privKey, pubKey, err := generateRSAKeys(4096)
		if err != nil {
			return err
		}
		now := time.Now()
		kp := &usertypes.KeyPair{
			Id:         uuid.New().String(),
			CreatedAt:  now,
			UpdatedAt:  now,
			PublicKey:  pubKey,
			PrivateKey: privKey,
		}
		err = s.keyPairsStore.Upsert(kp)
		if err != nil {
			return err
		}

		privKeyTyped, err := privateKeyFromString(kp.PrivateKey)
		if err != nil {
			return err
		}
		s.privateKey = privKeyTyped
		s.publicKeyPem = kp.PublicKey

	}

	// bootstrap admin user

	count, err := s.usersStore.Query(
		datastore.All(),
	).Count()

	if err != nil {
		return err
	}

	if count > 0 {
		return nil
	}

	_, err = s.register("singulatron", "changeme", "Admin", []string{
		usertypes.RoleAdmin.Id,
	})
	if err != nil {
		return err
	}

	// bootstrapping service user

	credentials, err := s.credentialsStore.Query(datastore.All()).Find()
	if err != nil {
		return err
	}

	email := "user-svc"
	pw := ""

	if len(credentials) > 0 {
		cred := credentials[0].(*usertypes.Credential)
		email = cred.Email
		pw = cred.Password
	} else {
		pw = uuid.New().String()
		err = s.credentialsStore.Upsert(&usertypes.Credential{
			Email:    email,
			Password: pw,
		})
		if err != nil {
			return err
		}
	}

	tok, err := s.login(email, pw)
	if err != nil {
		logger.Info(fmt.Sprintf("Registering the %v service", email))

		usr, err := s.register(email, pw,
			"User Service", []string{
				usertypes.RoleUser.Id,
			})
		if err != nil {
			return err
		}
		s.serviceUserId = usr.Id
	} else {
		usr, err := s.readUserByToken(tok.Token)
		if err != nil {
			return err
		}
		s.serviceUserId = usr.Id
	}

	return nil
}

func (s *UserService) registerRoles() error {
	_, err := s.UpsertRole(
		usertypes.RoleAdmin.Id,
		usertypes.RoleAdmin.Name,
		"",
		usertypes.RoleAdmin.PermissionIds,
	)
	if err != nil {
		return err
	}

	_, err = s.UpsertRole(
		usertypes.RoleUser.Id,
		usertypes.RoleUser.Name,
		"",
		usertypes.RoleUser.PermissionIds,
	)
	if err != nil {
		return err
	}

	return nil
}
